lib/core_ext/numpy.rb

```diff
@@ -59,6 +75,59 @@ def roll(shift, axis: 0)
     end
   end
 
+  # set the passed value to all cells in the specified region
+  def full(r1, c1, r2, c2, value)
+    pts!(r1, c1, r2, c2)
+    r1.upto(r2) { |r| c1.upto(c2) { |c| self[r][c] = value } }
+
+    self
+  end
+
+  # set the cells starting at the specified coordinates with the contents of the passed array
+  def paste(r, c, other)
+    if r < 0 || c < 0 || c + other.first.length >= first.length || r + other.length >= length
+      raise ArgumentError, 'must be in bounds'
+    end
+
+    other.each_with_index { |row, r2| row.each_index { |c2| self[r2 + r][c2 + c] = other[r2][c2] } }
+
+    self
+  end
+
+  # add the passed difference to all values in the specified region
+  def add(r1, c1, r2, c2, difference)
+    pts!(r1, c1, r2, c2)
+
+    r1.upto(r2) { |r| c1.upto(c2) { |c| self[r][c] += difference } }
+
+    self
+  end
+
+  # create new array from specified region
+  def copy(r1, c1, r2, c2)
+    pts!(r1, c1, r2, c2)
+
+    Array.new(r2 - r1 + 1).map { Array.new(c2 - c1 + 1) }.tap do |new_array|
+      r1.upto(r2) { |r| c1.upto(c2) { |c| new_array[r - r1][c - c1] = self[r][c] } }
+    end
+  end
+
+  # invert values in the specified region (true => false, false => true, nonzero => 0, and 0 => 1)
+  def not!(r1, c1, r2, c2)
+    pts!(r1, c1, r2, c2)
+
+    r1.upto(r2) do |r|
+      c1.upto(c2) do |c|
+        self[r][c] = case self[r][c]
+        when true then false
+        when false then true
+        when 0 then 1
+        else 0
```

Comment:
I think this should be a `when 1 then 0` and an `else` that raises "unsupported" or something